consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+	timeoutsCh   chan timeout
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+type state[V Hashable[H], H Hash] struct {
+	height uint
+	round  uint
+	step   step
+
+	lockedValue *V
+	validValue  *V
+
+	// The default value of lockedRound and validRound is -1. However, using int for one value is not good use of space,
+	// therefore, uint is used and nil would represent -1.
+	lockedRound *uint
+	validRound  *uint
+
+	// The following are round level variable therefore when a round changes they must be reset.
+	line34Executed bool
+	line36Executed bool
+	line47Executed bool
+}
+
+// Todo: Add Slashers later
+func New[V Hashable[H], H Hash, A Addr](addr A, app Application[V, H], chain Blockchain[V, H, A], vals Validators[A],
+	listeners Listeners[V, H, A], broadcasters Broadcasters[V, H, A], tmPropose, tmPrevote, tmPrecommit timeoutFn,
+) *Tendermint[V, H, A] {
+	return &Tendermint[V, H, A]{
+		nodeAddr:         addr,
+		state:            state[V, H]{height: chain.Height()},
+		messages:         newMessages[V, H, A](),
+		timeoutPropose:   tmPropose,
+		timeoutPrevote:   tmPrevote,
+		timeoutPrecommit: tmPrecommit,
+		application:      app,
+		blockchain:       chain,
+		validators:       vals,
+		listeners:        listeners,
+		broadcasters:     broadcasters,
+		scheduledTms:     make([]timeout, 0),
+		timeoutsCh:       make(chan timeout),
+		quit:             make(chan struct{}),
+	}
+}
+
+//nolint:gocyclo,funlen
+func (t *Tendermint[V, H, A]) Start() {
+	t.wg.Add(1)
+	go func() {
+		defer t.wg.Done()
+
+		t.startRound(0)
+
+		// Todo: check message signature everytime a message is received.
+		// For the time being it can be assumed the signature is correct.
+		i := 0
+		for {
+			fmt.Println("iteration: ", i)
+			select {
+			case <-t.quit:
+				fmt.Println("quit", t.state.step)
+				return
+			case p := <-t.listeners.ProposalListener.Listen():
+				if p.Height < t.state.height {
+					continue
+				}
+
+				if p.Height > t.state.height { //nolint:staticcheck
+					continue
+				}
+
+				if p.Round > t.state.round {
+					// Todo: add the message to a future round messages and then replay them after arriving in that
+					// round.
+					t.line55(p.Round)
+				}
+
+				// The below shouldn't panic because it is expected Proposal is well-formed. However,
+				// there need to be a way to distinguish between nil and zero value.
+				// This is expected to be handled by the p2p layer.
+				vID := (*p.Value).Hash()
+				fmt.Println("got proposal with valueID and validRound", vID, p.ValidRound)
+				validProposal := t.application.Valid(*p.Value)
+				proposalFromProposer := p.Sender == t.validators.Proposer(p.Height, p.Round)
+				vr := p.ValidRound
+
+				if validProposal {
+					// Add the proposal to the message set even if the sender is not the proposer,
+					// this is because of slahsing purposes
+					t.messages.addProposal(p)
+				}
+
+				_, prevotesForHR, precommitsForHR := t.messages.allMessages(p.Height, p.Round)
```

Comment:
How can there be prevotes and precommits at the proposal stage?? I thought the flow is prosoal -> prevote -> precommit..

Reply:
Ah, I didn't realise that proposers could use the prevotes (and precommits?) of previous rounds to help speed up the flow in the current round.